### ISSUE TITLE
How to vendor scripts in pathogen repos?

### ISSUE THREAD

**Issue by jameshadfield** [nextstrain/ingest#3]
This issue was originally written as an overview of `git subtree`, but later repurposed into a discussion of how to vendor scripts, specifically choosing between `git subtree` and `git subrepo`. In the end, we settled on `git subrepo` noting that this is a small implementation detail that can vary by pathogen repo, and be changed in the future.

## Original issue

This is a summary of how I used `git subtree` as part of #2. Note that different pipelines can choose different methods of vendoring scripts from this repo, but `git subtree` is particularly nice as it requires no knowledge of its existence from a user of the pipeline. 

Helpful reading: [Git Subtree basics](https://gist.github.com/SKempin/b7857a6ff6bddb05717cc17a44091202)

The script was added to this repo (nextstrain/ingest) from within nextstrain/hepatitisB using ingest as a subtree. Specifically, from the hepB repo:

```bash
# use a branch (in hepB)
git checkout -b 'vendored-scripts'

# add the ingest repo as a subtre, using the 'apply-geolocation-rules' branch
git remote add ingest-remote git@github.com:nextstrain/ingest.git
git subtree add --prefix ingest/vendored ingest-remote apply-geolocation-rules --squash
# Adds a merge commit with one parent the previous host repo HEAD commit,
#     and the other a squashed commit of the 'ingest' repo

# move the script to the subtree repo (ingest/vendored), modify the snakemake
# rules accordingly and commit changes (to the hepB repo)

# push the changes up to the subtree repo (ingest) on branch apply-geolocation-rules
git subtree push --prefix ingest/vendored ingest-remote apply-geolocation-rules
# The commit message was identical, but only the changes to ingest/vendored
# were part of the subtree commit (probably obvious!)
```

It was tested in monkeypox by pulling in (this branch of) the ingest repo as a subtree, and updating the transform rule accordingly. 

### Reflections

This approach is pretty straightforward but changing the branch of a subtree seems [to pollute the git history a bit](https://stackoverflow.com/a/18537904). An alternative approach would be to simply have a subtree of ingest at the `main` branch, [push any changes to the subtree up to a branch of the subtree](https://stackoverflow.com/questions/40003831/push-changes-from-git-subtree-to-a-branch-for-a-pull-request), merge that branch via GitHub (with code review etc), then pull down the changes once they're on the `main` branch (of the subtree repo). 

Given a script with differences in multiple repos, the most straightforward may be to simply to create a to-be-vendored version of a script locally, copy it into each repo to test, and when you are satisfied create a PR in this `ingest` repo without using subtrees at all. Once it's in `main`, it is straightforward to pull it into each pathogen repo using `git subtree pull ...`.

Comments / improvements welcome. At the very least this may give others a quick start guide!

**Comment by victorlin** [comment]
What's the use case of changing the branch of a subtree? Always pulling from `nextstrain/ingest@main` seems like a straightforward approach, and mimics how things work with other shared scripts in `augur curate`.

**Comment by joverlee521** [comment]
I briefly explored [git-subrepo](https://github.com/ingydotnet/git-subrepo) as a potential alternative to `git subtree`.

Some things that are nice compared to `git subtree`:
- It's obvious that a directory is a subrepo because it contains an auto-generated `.gitrepo` file
- The `.gitrepo` file contains info on the remote source that the current subrepo is using.
- Switching branches in the subrepo is painless and creates a single commit with message that summarizes the switch

I tested this out by creating a subrepo for nextstrain/ingest in my personal testing repo. 
These [two commits](https://github.com/joverlee521/nextstrain-testing/compare/0ba7fdd482fff0b8ea6f38af87bd2382e4d1e2d4...93f9c4126d7301e3429ad1d93b08273ead87fb6c) include the initial clone of the subrepo and a switch to the `apply-geolocation-rules` branch.

**Comment by jameshadfield** [comment]
> What's the use case of changing the branch of a subtree? Always pulling from nextstrain/ingest@main seems like a straightforward approach, and mimics how things work with other shared scripts in augur curate.

One use case is when a script needs modifications and you want to push them, however I think my subsequently suggested way to do this outside of `subtree` (or `subrepo`) is better. Hopefully we don't have too many occurrences of these. 

git-subrepo looks good too. I'll give it a try. It shares the very nice feature that a user of the pipeline (and oftentimes the developer) doesn't need to know anything about it!

**Comment by victorlin** [comment]
I tried this myself for the same `apply-geolocation-rules` script. See the following PRs:

- https://github.com/nextstrain/ingest/pull/4
- https://github.com/nextstrain/ingest/pull/5
- https://github.com/nextstrain/hepatitisB/pull/3
- https://github.com/nextstrain/hepatitisB/pull/4

Notes:

- I did not use `git subtree push`. A summary of my approach can be found at https://github.com/nextstrain/hepatitisB/commit/c8c94af059fbac72e18c6d20aa575906ae255904.
- I did not use `--squash` since it can be useful to see the commit messages from nextstrain/ingest in pathogen repos.

**Comment by victorlin** [comment]
Example PRs using `git subrepo`:

- https://github.com/nextstrain/hepatitisB/pull/5
- https://github.com/nextstrain/hepatitisB/pull/6

**Comment by victorlin** [comment]
Thoughts on `subtree` vs. `subrepo`:

- `subrepo` does not provide a way to preserve commit messages from the other repo, as it sees that feature as a problem. This makes it comparable to using `subtree` with `--squash`. Between those two, `subrepo` has cleaner commit history:
    - `subrepo` makes 1 squash commit based on the history of the pathogen repo
    - `subtree --squash` makes 1 squash commit based on external history + 1 merge commit to bring it back into the pathogen repo's history
- `subrepo` requires extra installation for anyone who wants to set up usage of centralized scripts (`subtree clone`) or pull new changes `subtree pull`.
- With a pull-only approach, there is flexibility: individual pathogen repo maintainers can choose between `subtree` or `subrepo`. Of course, we can provide suggestions here.
- `subrepo`'s README lists several benefits compared to the built-in git commands. It's not yet clear to me which of these could be downsides of sticking with `subtree`, but it seems most of them can be avoided by the pull-only approach.

**Comment by joverlee521** [comment]
We are going with `git subtree`, based on internal team discussion on [Slack](https://bedfordlab.slack.com/archives/C01LCTT7JNN/p1690504459388819).

Added vendoring section to the README in https://github.com/nextstrain/ingest/commit/87ca1edc779303b2c2fd80541e357aeec0ceea71.

**Comment by victorlin** [comment]
From https://github.com/nextstrain/ingest/issues/3#issuecomment-1633248200:

> I did not use `--squash` since it can be useful to see the commit messages from nextstrain/ingest in pathogen repos.

I just realized that these commit messages aren't visible when looking at the log of a particular file, but they are in the blame. Examples:

- `subtree` [log](https://github.com/nextstrain/hepatitisB/commits/victorlin/use-centralized-apply-geolocation-rules/ingest/vendored/apply-geolocation-rules)
- `subtree` [blame](https://github.com/nextstrain/hepatitisB/blame/victorlin/use-centralized-apply-geolocation-rules/ingest/vendored/apply-geolocation-rules)
- `subrepo` [log](https://github.com/nextstrain/hepatitisB/commits/victorlin/use-centralized-apply-geolocation-rules-git-subrepo/ingest/vendored/apply-geolocation-rules)
- `subrepo` [blame](https://github.com/nextstrain/hepatitisB/blame/victorlin/use-centralized-apply-geolocation-rules-git-subrepo/ingest/vendored/apply-geolocation-rules)

Having the blame might be useful, at the cost of a more verbose git history. I think the only way to truly understand the value here is to use it in practice for a bit!

**Comment by joverlee521** [comment]
Oh, just figured out it's possible to see the commit history of a subrepo (without all of the git history pollution):
```
git subrepo fetch ingest/vendored
git log refs/subrepo/ingest/vendored/fetch 
```

You can also see commit history of a single file if you provide the filepath from the ingest repo:
```
git log refs/subrepo/ingest/vendored/fetch -- notify-slack
```

**Comment by victorlin** [comment]
I'm looking back at the [Slack conversation](https://bedfordlab.slack.com/archives/C01LCTT7JNN/p1690504459388819) and it seems like we were on the fence about `subrepo` vs. `subtree` and I chose one mostly to have a decision to move forward.

Since we've adopted `git subrepo`, we've ran into some issues such as #21, and recently @tsibley made a "[bad suggestion](https://github.com/nextstrain/docs.nextstrain.org/issues/179#issuecomment-1911149718)" (which I don't think is that bad) to use `git subtree`. I think it can be reconsidered. If other people think so too, I'll make PRs to switch to `git subtree` in the pathogen repos, and we can see how it looks.

Things to consider:

- We've already spent some time with `git subrepo`, and it works well enough. `git subtree` might come with new hurdles that we haven't discovered yet.
- It's meant to be a temporarily solution, though external usage is likely with https://github.com/nextstrain/docs.nextstrain.org/issues/179.
- In the end, the difference for external usage is just an extra installation step for `git subrepo` which is straightforward.
